### ISSUE TITLE
Support groupings for plugins

### ISSUE THREAD

**Issue by adamltyson** [napari/napari#2394]
## 🚀 Feature
It would be great if plugin developers could provide some kind of grouping for our plugins.

## Motivation
Gradually I plan to provide napari plugins for most of the tools I maintain, and this will likely end up being >10 analysis and IO plugins. These could get messy and disorganised quickly (they already are). 

## Pitch
It would be cool if plugins in multiple repos could be organised in some kind of hierarchy e.g. `Plugins` -> `Add dock widget` -> `Organisation` -> `Tool` -> `Widget`. Even better if these groups can be used in the list of available plugins. 

## Alternatives

I can provide a single `organisation-napari` package that provides widgets from N package. This has two problems:
* Limited organisation (One level of hierarchy)
* Causes dependency issues. I don't want to install tensorflow when users just wanted an IO plugin.

## Additional context

Long term, this could help with plugin discoverability. Without some kind of organisation, in a few years there could be hundreds of plugins. It's very easy to add the trove classifier and clutter up the list.

**Comment by sofroniewn** [comment]
Very interesting ideas @adamltyson, definitely worth discussing more at some of our developer meetings. I don't have any immediate answers, but we'll need to think and discuss this more! Curious what others could imagine cc @jni @tlambert03 @neuromusic

**Comment by neuromusic** [comment]
very cool idea.

there's been some discussion of "bundles" of plugins... for example, a "cell bio starter pack" or "so-and-so's lab's plugins". there are [examples of this in VS Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack). there are a few ways that could be enabled to address the discoverability issue, but you bring up a really good point that it will be important that such context isn't lost once the user has installed the plugin and is looking for it in napari.

ultimately, as the number of plugins someone has installed grows, any hierarchy becomes limiting... there are many different kinds of hierarchies/ontologies for organizing plugins which optimize for different use cases. this is probably why atom/vscode have opted for search/autocomplete to activate plugins and their functions.

this is a rambly way of saying... yes, I 100% agree that it will be invaluable to let plugin devs define some way that a set of plugins "belongs" to the dev/org

**Comment by jni** [comment]
I agree that this is something we'll have to tackle eventually, but we need some way of handling "authorisation" for the bundles, so to speak. What's to stop any Joe off the street from claiming their plugin belongs in the Cellfinder org? =)

**Comment by adamltyson** [comment]
> ultimately, as the number of plugins someone has installed grows, any hierarchy becomes limiting... there are many different kinds of hierarchies/ontologies for organizing plugins which optimize for different use cases. this is probably why atom/vscode have opted for search/autocomplete to activate plugins and their functions.

Would it be possible for developers to suggest a hierarchy for their plugin? e.g.:
```python
LEVEL_0 = "BrainGlobe"
LEVEL_1 = "cellfinder"
```

But then then the user could rearrange the plugins to suit them, and these preferences could be saved, and would superseed the developers' suggested hierarchy. That way, we can distribute them in an organised way for discoverability, but the user is free to create their own folders of plugins?

I'm not sure how to handle bundles of plugins, if these bundles are created by a third party, but the priority could be user > bundle > package default.

----

>  but we need some way of handling "authorisation" for the bundles,

Has there been any discussion about handling this generally? This seems like a difficult problem, even now, there's nothing stopping someone uploading a random package called `cellfinderV2` . There could be no "authorisation", but a mechanism (e.g. issues on a specific repository)  to allow requests for packages to be added to a block list. This could be malicious code, packages with some naming issue (offensive, or impersonating others), or even if a developer just wanted an old package to no longer be listed.

**Comment by jni** [comment]
> This seems like a difficult problem, even now, there's nothing stopping someone uploading a random package called cellfinderV2

You're spot on, of course. I dunno if you saw [this little episode from last week](https://github.com/pypa/pypi-support/issues/923#issuecomment-787613805). :joy: :grimacing:
